### PR TITLE
unmodifiableList vs. List.of - explanation & example

### PR DESCRIPTION
**Pull request by Nimrud**

### CODE REVIEW

**Review by asilenko — COMMENTED** [review]
Nice examples and impressive delivery time.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+package com.epam.prd.jjop.helpers.unmodifiableList_vs_List_of;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The purpose of this class is to show the difference between:
+ * => .unmodifiableList()
+ * => List.of()
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java:10]:
```suggestion
 * <ul>
 *   <li>{@link java.util.Collections#unmodifiableList(List<? extends T>)</li>
 *   <li>java.util.List#of()</li>
 * </ul>
```

You may link directly to proper classes and their members via `{@link ...}` in javadocs. I may slightly mistake the syntax but it is close to suggestion above.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    public static void main(String[] args) {
+
+        /**
+         * SUMMARY of the difference between .unmodifiableList() vs List.of():
+         * while we cannot perform any operations on those 'types' of lists,
+         * contents of List.of() never change, while the values of .unmodifiableList() can change
+         */
+
+        /**
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java:27]: Don't divide the javadocs into separate blocks. Connect those into one.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+         * - don't want to expose the original list to any other classes - that way we do not use the original list,
+         * but a copy (original list is encapsulated).
+         * This should display the exact same array to the console.
+         */
+        System.out.println("list (unmodifiableList) = " + getList(list));
+
+        /**
+         * Step 3: We modify the original array: set first element to 4 and second to 25
+         * (uncomment to run)
+         */
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/Main.java:51]:
```suggestion
         */
         TODO: uncomment the code below.
```

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/example/SomeOtherClass.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Please read contents of the Main class first
+ * @see com.epam.prd.jjop.helpers.unmodifiableList_vs_List_of.Main
+ */
+
+class SomeOtherClass {
+    public static void main(String[] args) {
+
+        final Basket basket = new Basket();
+        /**
+         * retrieving the list from a different class:
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiableList_vs_List_of/example/SomeOtherClass.java:15]:
```suggestion
         * Retrieving the list from a different class:
```

**Comment by marcin-ogorzalek-EPAM** [comment]
Generally nice work, you showed the main difference between List.of() and Collections.unmodifaebleList().

I have 3 requests:

* improve your javadocs (those should follow conventions more closely, more tags, proper English sentences, HTML tags if necessary)
* more TOD comments (especially near commented code)
* move two additional classes you created to one file (as there can be more than one class in one file) and document Basket with at least class javadoc.

**Comment by Nimrud** [comment]
Thank you all for your comments! Will update my solution to include my suggestions.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ *     <li>{@link List#of()}</li>
+ * </ul>
+ *
+ * @author Krzysztof Jaczewski
+ */
+
+public class Main {
+
+    private static List<Integer> list;
+    private static List<Integer> list_second;
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:20]: name breaks Java nem,ing convention

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
+package com.epam.prd.jjop.helpers.unmodifiablelist_vs_list_of;
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:1]: By convention, packages are named without underscores. All words should be concatenated. I know that `unmodifiablelistvslistof` doesn't looks but regarding to the convention is should be like this.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The purpose of this class is to show the difference between:
+ * <ul>
+ *     <li>{@link java.util.Collections#unmodifiableList(List)}</li>
+ *     <li>{@link List#of()}</li>
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:11]: Is this working with CTRL+B click or requires FQN?

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+         *     <li>can't make any operation on unmodifiableList (e.g. add anything) but the list itself CAN change;
+         *     example at:
+         *     @see com.epam.prd.jjop.helpers.unmodifiablelist_vs_list_of.example.SomeOtherClass</li>
+         * </ul>
+         *
+         *
+         *
+         * ===> List.of <===
+         *
+         * List.of() creates a list that cannot change.
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:89]: why no `{@link}` here?

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ * The purpose of this class is to show the difference between:
+ * <ul>
+ *     <li>{@link java.util.Collections#unmodifiableList(List)}</li>
+ *     <li>{@link List#of()}</li>
+ * </ul>
+ *
+ * @author Krzysztof Jaczewski
+ */
+
+public class Main {
```
Comment (marcin-ogorzalek-EPAM) [inline src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:17]: why so public?

**Comment by marcin-ogorzalek-EPAM** [comment]
Definitely, there can be more `{@link}`, `{@code}` and other javadocs improvements introduced (not only those hinted by my).

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java (hunk 5 above)
Comment (Nimrud) [reply src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:20]: Yeah, stupid me, did not catch it.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java (hunk 7 above)
Comment (Nimrud) [reply src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:11]: That's what was automatically converted into by Intellij. I typed in FQN.

File: src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java (hunk 8 above)
Comment (Nimrud) [reply src/main/java/com/epam/prd/jjop/helpers/unmodifiablelist_vs_list_of/Main.java:89]: I have already used a linked version on top of the class. Not sure if you want me to change all occurrences of List.of() in the class to a linked version or just here.

**Pull request merged**